src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -124,13 +138,13 @@ public void Reset()
             Debug.Assert(IsCompleted);
             Debug.Assert(_callbacks == null);
             _isCompleted = false;
-            _exceptionInfo = null;
+            _exception = null;
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
         private void ThrowLatchedException()
         {
-            _exceptionInfo.Throw();
+            // Throw a new exception so as not to corrupt stack trace, as may be thrown multiple times
+            throw new IOException(_exception.Message, _exception);
```

Comment:
Does anyone depend on the actual type of _exception? e.g. could it be an OperationCanceledException that someone might be paying attention to for a signal of cancellation?

Reply:
Tasks and async :)

Will add a pattern match

Reply:
Yes this seems very wrong. Wrapping every exception in an IO exception will break tons of code.